### PR TITLE
add ability to post-process earlier images using the CLI

### PR DESCRIPTION
**Pull request by lstein**
- Supports gfpgan, esrgan, codeformer and embiggen
- To use: 

~~~~
dream> !fix ./outputs/img-samples/000056.292144555.png -ft gfpgan -U2 -G0.8 
dream> !fix ./outputs/img-samples/000056.292144555.png -ft codeformer -G 0.8 
dream> !fix ./outputs/img-samples/000056.29214455.png  -U4 
dream> !fix ./outputs/img-samples/000056.292144555.png -embiggen 1.5
~~~~
- The first example invokes gfpgan to fix faces and esrgan to upscale.
- The second example invokes codeformer to fix faces, no upscaling
- The third example uses esrgan to upscale 4X
- The four example runs embiggen to enlarge 1.5X

You are also supposed to be able to provide just the basename of the file, and it will automatically prepend the current outdir path, but I forgot to test that functionality and have to break for the evening.

- This is very preliminary work. There are some anomalies to note:
  1. The syntax is non-obvious. I would prefer something like: !fix esrgan,gfpgan !fix esrgan !fix embiggen,codeformer

     However, this will require refactoring the gfpgan and embiggen code.

   2. Images generated using gfpgan, esrgan or codeformer all are named "xxxxxx.xxxxxx.postprocessed.png" and the original is saved. However, the prefix is a new one that is not related to the original.

   3. Images generated using embiggen are named "xxxxx.xxxxxxx.png", and once again the prefix is new. I'm not sure whether the prefix should be aligned with the original file's prefix or not. Probably not, but opinions welcome.

### CODE REVIEW

**Comment by blessedcoolant** [comment]
Oh thank you so much for working on this. Has been on list of to do things too. I was waiting for your check on this #547 before I did.

I'd still ask you to have a look at it. I refactored the implementation of both GFPGAN and ESRGAN to be independent. They follow the same format as the new CodeFormer code. Maybe we can merge it in with this so it's all consistent?

**Comment by lstein** [comment]
So sorry! I had started down the path of a queuing system for postprocessor steps and thought that this would require more refactoring, so I put your pull request on hold. Then @Kyle0654 introduced his new dependency-injector framework, and I dropped my plans for the queuing. I'll review PR #547 tonight and merge it in.

**Comment by lstein** [comment]
Will wait until @blessedcoolant completes work on pr #547.

**Comment by lstein** [comment]
@blessedcoolant I've rebased your PR #690 into this and tweaked a few things to get the changes to cooperate. Could you kick the tires and let me know if we're good to go? I'll keep PR #690 open until you confirm.

**Comment by blessedcoolant** [comment]
> @blessedcoolant I've rebased your PR #690 into this and tweaked a few things to get the changes to cooperate. Could you kick the tires and let me know if we're good to go? I'll keep PR #690 open until you confirm.

I'll test this out in a few. Just refining some stuff on the progress feature at the moment.

**Comment by lstein** [comment]
Just a nudge to review this. I'm reviewing your in-progress image work now.

File: docs/features/UPSCALE.md [hunk 1]
```diff
@@ -4,14 +4,16 @@ title: Upscale
...
-The script provides the ability to restore faces and upscale.
+The script provides the ability to restore faces and upscale. You can apply these operations
+at the time you generate the images, or at any time to a previously-generated PNG file, using 
+the [!fix](#Fixing Previously-Generated Images) command.
 
-You can enable these features by passing `--restore` and `--esrgan` to your launch script to enable
-face restoration modules and upscaling modules respectively.
+# :material-image-size-select-large: Upscale 
 
-## **GFPGAN and Real-ESRGAN Support**
```
Comment (tildebyte) [inline docs/features/UPSCALE.md:7]: Don't bold headings. `##` is plenty.

File: docs/features/UPSCALE.md [hunk 2]
```diff
@@ -4,14 +4,16 @@ title: Upscale
 
 ## **Intro**
```
Comment (tildebyte) [inline docs/features/UPSCALE.md:5]: Don't bold headings. `##` is plenty.

**Pull request merged**